driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -2735,9 +2735,19 @@ public void handle(Message.Response response) {
         case TOPOLOGY_CHANGE:
           ProtocolEvent.TopologyChange tpc = (ProtocolEvent.TopologyChange) event;
           Host.statesLogger.debug("[{}] received event {}", tpc.node, tpc.change);
-          // Always do a full refresh for topology changes. This is simpler than trying to infer a
-          // new EndPoint from the rpc_address only.
-          submitNodeListRefresh();
+          // Do NOT translate the address, it will be matched against Host.getBroadcastRpcAddress()
+          // to find the target host.
+          switch (tpc.change) {
+            case NEW_NODE:
+              submitNodeRefresh(tpc.node, HostEvent.ADDED);
+              break;
+            case REMOVED_NODE:
+              submitNodeRefresh(tpc.node, HostEvent.REMOVED);
+              break;
+            case MOVED_NODE:
```

Comment:
Side note: I'm not sure where this MOVED_NODE event comes from, it's not referenced anywhere in the protocol spec.
It's been in the codebase for a long time, maybe it used to be a thing.

Driver 4 does not have it (nor do I think it should).